### PR TITLE
feat: add endpoints for networks

### PR DESCRIPTION
**Pull request by jungwire**
### Description:

Adds REST API endpoints for the networks. 
Exposes:
 - GET /networks
 - GET /networks/{id}
 - GET /networks/{id}/negotiations
 - GET /networks/{id}/managers
 - GET /networks/{id}/resources

### Checklist:

_Make sure you tick all the boxes bellow if they are true or do not apply:_

- [x] I have performed a self review of my code
- [x] My code follows [Google Java Code style](https://google.github.io/styleguide/javaguide.html)
- [x] I have made my code as simple as possible
- [x] I have added unit tests and the code coverage has not decreased
- [ ] I have updated the documentation in all relevant places

### CODE REVIEW

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.springframework.hateoas.PagedModel;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/v3")
+@Tag(name = "Networks", description = "Retrieve connected networks")
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:29]: It should not be only retrieval, Please add endpoints for delete, update as well

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  public PagedModel<EntityModel<NetworkDTO>> list(
+      @RequestParam(required = false, defaultValue = "0") int page,
+      @RequestParam(required = false, defaultValue = "50") int size) {
+    return networkModelAssembler.toPagedModel(
+        (Page<NetworkDTO>) networkService.findAllNetworks(PageRequest.of(page, size)));
+  }
+
+  @GetMapping("/networks/{id}")
+  @Operation(summary = "Get network by id")
+  public NetworkDTO findById(@PathVariable("id") Long id) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:67]: this should be an EntityModel as well

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      @PathVariable("id") Long id,
+      @RequestParam(required = false, defaultValue = "0") int page,
+      @RequestParam(required = false, defaultValue = "50") int size) {
+    return resourceModelAssembler.toPagedModel(
+        (Page<ResourceResponseModel>)
+            resourceService.findAllForNetwork(PageRequest.of(page, size), id));
+  }
+
+  @GetMapping("/networks/{id}/managers")
+  @Operation(summary = "List all managers of a network")
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:83]: Paged model

File: src/main/java/eu/bbmri_eric/negotiator/configuration/security/OAuthSecurityConfig.java [hunk 4]
```diff
@@ -220,6 +220,6 @@ public OAuth2TokenValidator<Jwt> introspectionValidator() {
   }
 
   OAuth2TokenValidator<Jwt> audienceValidator() {
-    return new JwtClaimValidator<List<String>>(AUD, aud -> aud.contains(audience));
+    return new JwtClaimValidator<List<String>>(AUD, aud -> aud != null && aud.contains(audience));
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/configuration/security/OAuthSecurityConfig.java:223]: why?

File: src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java [hunk 5]
```diff
@@ -1,134 +1,99 @@
...
-/**
- * Represents a Network entity in the database. A Network has a unique identifier, a URI, a name, a
- * contact email, a set of managers, and a set of resources.
- */
 @Getter
 @Setter
 @Entity
 @Builder
-@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java:14]: Why not protected?

Comment (RadovanTomik) [reply src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java:14]: please rebase this to latest master

Comment (RadovanTomik) [reply src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java:14]: I think you have mix of old and new commits, please fix it

File: src/main/java/eu/bbmri_eric/negotiator/mappers/NegotiationModelAssembler.java [hunk 6]
```diff
@@ -68,6 +68,14 @@ public PagedModel<EntityModel<NegotiationDTO>> toPagedModel(
         links);
   }
 
+  public PagedModel<EntityModel<NegotiationDTO>> toPagedModel(@NonNull Page<NegotiationDTO> page) {
+    return PagedModel.of(
+        page.getContent().stream().map(this::toModel).collect(Collectors.toList()),
+        new PagedModel.PageMetadata(
+            page.getSize(), page.getNumber(), page.getTotalElements(), page.getTotalPages()),
+        new ArrayList<>());
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/mappers/NegotiationModelAssembler.java:76]: add links

File: src/main/java/eu/bbmri_eric/negotiator/mappers/NetworkModelAssembler.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                .withSelfRel(),
+            WebMvcLinkBuilder.linkTo(NetworkController.class).withRel("networks"));
+  }
+
+  public PagedModel<EntityModel<NetworkDTO>> toPagedModel(@NonNull Page<NetworkDTO> page) {
+    return PagedModel.of(
+        page.getContent().stream().map(this::toModel).collect(Collectors.toList()),
+        new PagedModel.PageMetadata(
+            page.getSize(), page.getNumber(), page.getTotalElements(), page.getTotalPages()),
+        new ArrayList<>());
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/mappers/NetworkModelAssembler.java:34]: add links

File: src/main/java/eu/bbmri_eric/negotiator/service/NetworkService.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package eu.bbmri_eric.negotiator.service;
+
+import eu.bbmri_eric.negotiator.dto.NetworkDTO;
+import org.springframework.data.domain.Pageable;
+
+public interface NetworkService {
+
+  /**
+   * Retrieves a NetworkDTO with the specified ID.
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/NetworkService.java:9]: not sure about mentioning NetwrokDTP

File: src/test/java/eu/bbmri_eric/negotiator/integration/repository/NetworkRepositoryTest.java [hunk 9]
```diff
@@ -24,15 +23,7 @@ public class NetworkRepositoryTest {
...
 
-  @Autowired ResourceRepository resourceRepository;
-
-  @Autowired DiscoveryServiceRepository discoveryServiceRepository;
-
-  @Autowired OrganizationRepository organizationRepository;
-  @Autowired private PersonRepository personRepository;
-
   @Test
-  @Transactional
```
Comment (RadovanTomik) [inline src/test/java/eu/bbmri_eric/negotiator/integration/repository/NetworkRepositoryTest.java:35]: Why is this deleted?

File: src/test/java/eu/bbmri_eric/negotiator/unit/model/NetworkTest.java [hunk 10]
```diff
@@ -31,11 +31,8 @@ void equals_differentId_notEqual() {
   @Test
   void addCollection() {
```
Comment (RadovanTomik) [inline src/test/java/eu/bbmri_eric/negotiator/unit/model/NetworkTest.java:32]: why is this deleted?

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
please see comments but mainly fix the outdated commits

File: src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java (hunk 5 above)
Comment (jungwire) [reply src/main/java/eu/bbmri_eric/negotiator/database/model/Network.java]: _"The JPA specification requires all Entity classes to have a default no-arg constructor. This can be either public or protected."_

https://www.baeldung.com/jpa-no-argument-constructor-entity-class

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  @PostMapping(
+      value = "/networks",
+      consumes = MediaType.APPLICATION_JSON_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  @Operation(summary = "Create a new network")
+  @ResponseStatus(HttpStatus.CREATED)
+  public EntityModel<NetworkDTO> create(
+      @Validated(ValidationGroups.Create.class) @RequestBody NetworkDTO networkDTO) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:91]: what happens if you post a NetworkDTO with an ID?

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+    return networkModelAssembler.toModel(networkService.createNetwork(networkDTO));
+  }
+
+  @GetMapping("/networks/{id}")
+  @Operation(summary = "Get network by id")
+  public EntityModel<NetworkDTO> findById(@PathVariable("id") Long id) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:97]: are you sure the ("id") is necessary?

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  @PutMapping(
+      value = "/networks/{id}",
+      consumes = MediaType.APPLICATION_JSON_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  @Operation(summary = "Update network by id")
+  public EntityModel<NetworkDTO> update(
+      @PathVariable("id") Long id,
+      @Validated(ValidationGroups.Update.class) @RequestBody NetworkDTO networkDTO) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:115]: what happens if you provide 2 different IDs?

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+  @DeleteMapping("/networks/{id}/managers/{managerId}")
+  @Operation(summary = "Remove manager from network")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void removeManagerFromNetwork(
+      @PathVariable("id") Long id, @PathVariable("managerId") Long managerId) {
+    networkService.removeManagerFromNetwork(id, managerId);
+  }
+
+  @GetMapping("/networks/{id}/negotiations")
+  @Operation(summary = "List all negotiations associated a network")
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:181]: please fix the grammar

File: src/main/java/eu/bbmri_eric/negotiator/mappers/UserModelAssembler.java [hunk 15]
```diff
@@ -158,4 +172,44 @@ private static List<Link> getLinks(Page<UserResponseModel> page) {
...
+              .withRel(IanaLinkRelations.PREVIOUS)
+              .expand(parameters));
+    }
+    if (page.hasNext()) {
+      links.add(
+          linkTo(
+                  methodOn(NetworkController.class)
+                      .getManagers(networkId, page.getNumber() + 1, page.getSize()))
+              .withRel(IanaLinkRelations.NEXT)
+              .expand(parameters));
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/mappers/UserModelAssembler.java:195]: are these links necessary?

File: src/main/java/eu/bbmri_eric/negotiator/service/NetworkService.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+   */
+  void addManagersToNetwork(Long networkId, List<Long> managerIds);
+
+  /**
+   * Adds a list of resources to a network.
+   *
+   * @param networkId
+   * @param resourceIds
+   */
+  void addResourcesToNetwork(Long networkId, List<Long> resourceIds);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/NetworkService.java:78]: please group similar methods together

File: src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  @Override
+  public void deleteNetworkById(Long id) {
+    Network network = getNetwork(id);
+    Set<Resource> resources = new HashSet<>(network.getResources());
+    Set<Person> managers = new HashSet<>(network.getManagers());
+    resources.forEach(resource -> network.removeResource(resource));
+    managers.forEach(manager -> network.removeManager(manager));
+    networkRepository.deleteById(id);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java:52]: can't this be handled by a JPA cascade annotation?

**Review by RadovanTomik — CHANGES_REQUESTED** [review]

File: src/main/java/eu/bbmri_eric/negotiator/configuration/BaseConfig.java [hunk 18]
```diff
@@ -20,7 +24,18 @@ public PasswordEncoder passwordEncoder() {
 
   @Bean
   public ModelMapper modelMapper() {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/configuration/BaseConfig.java:26]: why?

File: src/main/java/eu/bbmri_eric/negotiator/database/model/Organization.java [hunk 19]
```diff
@@ -1,16 +1,6 @@
...
-import jakarta.persistence.Entity;
-import jakarta.persistence.FetchType;
-import jakarta.persistence.GeneratedValue;
-import jakarta.persistence.GenerationType;
-import jakarta.persistence.Id;
-import jakarta.persistence.NamedAttributeNode;
-import jakarta.persistence.NamedEntityGraph;
-import jakarta.persistence.OneToMany;
-import jakarta.persistence.SequenceGenerator;
+import jakarta.persistence.*;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/model/Organization.java:3]: wildcard import

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java [hunk 20]
```diff
@@ -45,4 +47,14 @@ List<Negotiation> findByModifiedDateBeforeAndCurrentState(
...
+
+  @Query(
+      value =
+          "SELECT distinct (n) "
+              + "FROM Negotiation n "
+              + "JOIN n.requests rq "
+              + "JOIN rq.resources rs "
+              + "JOIN rs.networks net "
+              + "WHERE net.id = :networkId")
+  Page<Negotiation> findAllForNetwork(Long networkId, Pageable pageable);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java:59]: I do not think this will return just a page.

File: src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    return modelMapper.map(network, NetworkDTO.class);
+  }
+
+  @Override
+  public NetworkDTO createNetwork(NetworkCreateDTO networkCreateDTO)
+      throws EntityNotStorableException {
+    Network network = modelMapper.map(networkCreateDTO, Network.class);
+    try {
+      Network savedNetwork = networkRepository.saveAndFlush(network);
+      return modelMapper.map(savedNetwork, NetworkDTO.class);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java:83]: this is technically not correct. The entity could have been saved but the mapper returns an error. I would rather mark this as transactional and just save it. What exceptions do you want to handle?

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
please see comments

File: src/main/java/eu/bbmri_eric/negotiator/mappers/NetworkCreateModelMapper.java [hunk 22]
```diff
@@ -0,0 +1,27 @@
+package eu.bbmri_eric.negotiator.mappers;
+
+import eu.bbmri_eric.negotiator.database.model.Network;
+import eu.bbmri_eric.negotiator.dto.network.NetworkCreateDTO;
+import org.springframework.stereotype.Component;
+
+@Component
+public class NetworkCreateModelMapper {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/mappers/NetworkCreateModelMapper.java:8]: please look into how other mappers are implemented

File: src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+      throw new EntityNotStorableException("Database error occurred while saving network");
+    } catch (IllegalArgumentException iae) {
+      throw new EntityNotStorableException("Invalid data provided");
+    } catch (ConstraintViolationException cve) {
+      throw new EntityNotStorableException("Constraint violation");
+    } catch (TransactionException te) {
+      throw new EntityNotStorableException("Transaction error");
+    } catch (Exception ex) {
+      throw new EntityNotStorableException("An unexpected error occurred");
+    }
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/NetworkServiceImpl.java:100]: please handle these exceptions in the https://github.com/BBMRI-ERIC/negotiator/blob/5993c7b47e2ee48056e858aaeac1020e1f55ef9c/src/main/java/eu/bbmri_eric/negotiator/exceptions/NegotiatorExceptionHandler.java class. This is I think fine but it looks cluttered IMHO. because we would then need to do this for every single method that stores an entity,

**Review by RadovanTomik — CHANGES_REQUESTED** [review]

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+    networkService.removeResourceFromNetwork(id, resourceId);
+  }
+
+  @PostMapping(
+      value = "/networks/{id}/resources",
+      consumes = MediaType.APPLICATION_JSON_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  @Operation(summary = "Add a list of resources to a network")
+  @ResponseStatus(HttpStatus.OK)
+  public void addResourcesToNetwork(@PathVariable Long id, @RequestBody List<Long> resourceIds) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/NetworkController.java:146]: same as bellow

**Review by RadovanTomik — CHANGES_REQUESTED** [review]

**Review by RadovanTomik — APPROVED** [review]
